### PR TITLE
Free memory leaked by by cgi-bin

### PR DESCRIPTION
**Pull request by AtariDreams**
I am not going to sugar-coat this: there are potentially hundreds of tiny memory leaks.

The culprit is cgiGetVariable strdup the result, which allocates onto the heap, but many of these function calls assume that does not happen.

So far, I am taking the approach of freeing when done. However, given how much more messy that makes the code, I may just go along with returning a read-only pointer to the array itself, while not allowing any changes to happen.

I already made a function that checks if the result exists or is NULL without strdup, but this only checks if the variable exists.

What do you think?

### CODE REVIEW

**Comment by AtariDreams** [comment]
@zdohnal Any advice?

**Comment by michaelrsweet** [comment]
Keep in mind that the CGI programs have a very short lifetime so any “leaks” of form variables are resolved quickly when the corresponding process exits.

**Comment by AtariDreams** [comment]
> Keep in mind that the CGI programs have a very short lifetime so any “leaks” of form variables are resolved quickly when the corresponding process exits.

Ok so they don't continuously run in the background like daemons tend to do, right?

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
Remove unrelated changes.

**Comment by michaelrsweet** [comment]
> Ok so they don't continuously run in the background like daemons tend to do, right?

Correct, they run and go away.

File: cgi-bin/printers.c [hunk 1]
```diff
@@ -180,6 +180,7 @@ main(void)
       cgiCopyTemplateLang("error-op.tmpl");
       cgiEndHTML();
     }
+    free((void *)op);
```
Comment (michaelrsweet) [inline cgi-bin/printers.c:183]: why

File: cgi-bin/printers.c [hunk 2]
```diff
@@ -190,6 +191,7 @@ main(void)
     cgiStartHTML(cgiText(_("Printers")));
     cgiCopyTemplateLang("error-op.tmpl");
     cgiEndHTML();
+    free((void *)op);
```
Comment (michaelrsweet) [inline cgi-bin/printers.c:194]: why

File: cgi-bin/printers.c [hunk 3]
```diff
@@ -201,6 +203,7 @@ main(void)
  /*
   * Return with no errors...
   */
+  free((void *)op);
```
Comment (michaelrsweet) [inline cgi-bin/printers.c:206]: why

File: cgi-bin/printers.c [hunk 4]
```diff
@@ -350,9 +353,15 @@ show_all_printers(http_t     *http,	/* I - Connection to server */
...
-        !cgiGetVariable("CLEAR"))
-      search = cgiCompileSearch(var);
+    if ((var = cgiGetVariable("QUERY")) != NULL)
+    {
+      if (!cgiVariableExists("CLEAR"))
+        search = cgiCompileSearch(var);
+      else
+        search = NULL;
+      
+      free((void *)var);
```
Comment (michaelrsweet) [inline cgi-bin/printers.c:363]: why the free and why the cast?

File: cgi-bin/printers.c [hunk 5]
```diff
@@ -367,7 +376,10 @@ show_all_printers(http_t     *http,	/* I - Connection to server */
     */
 
     if ((var = cgiGetVariable("FIRST")) != NULL)
+    {
       first = atoi(var);
+      free((void *)var);
```
Comment (michaelrsweet) [inline cgi-bin/printers.c:381]: why

File: cgi-bin/search.c [hunk 6]
```diff
@@ -62,7 +62,7 @@ cgiCompileSearch(const char *query)	/* I - Query string */
 
   if ((s = (char *)malloc(slen)) == NULL)
   {
-    free(re);
+    free((void *)re);
```
Comment (michaelrsweet) [inline cgi-bin/search.c:65]: why the cast?

File: cgi-bin/search.c [hunk 7]
```diff
@@ -107,11 +107,11 @@ cgiCompileSearch(const char *query)	/* I - Query string */
         * No closing quote, error out!
 	*/
 
-	free(s);
-	free(re);
+	free((void *)s);
```
Comment (michaelrsweet) [inline cgi-bin/search.c:110]: ditto

File: cgi-bin/search.c [hunk 8]
```diff
@@ -178,11 +178,11 @@ cgiCompileSearch(const char *query)	/* I - Query string */
         temp = (char *)realloc(s, slen);
 	if (!temp)
 	{
-	  free(s);
-	  free(re);
+	  free((void *)s);
```
Comment (michaelrsweet) [inline cgi-bin/search.c:181]: ditto

File: cgi-bin/var.c [hunk 9]
```diff
@@ -234,6 +237,50 @@ cgiGetVariable(const char *name)	/* I - Name of variable */
   return ((var == NULL) ? NULL : strdup(var->values[var->nvalues - 1]));
 }
 
+/*
```
Comment (michaelrsweet) [inline cgi-bin/var.c:240]: two blank lines

File: cgi-bin/var.c [hunk 10]
```diff
@@ -234,6 +237,50 @@ cgiGetVariable(const char *name)	/* I - Name of variable */
...
+cgiVariableExists(const char *name)	/* I - Name of variable */
+{
+  const _cgi_var_t	*var;		/* Returned variable */
+
+
+  var = cgi_find_variable(name);
+
+  return (var && var->values[var->nvalues - 1]);
+}
+
```
Comment (michaelrsweet) [inline cgi-bin/var.c:256]: two blank lines

File: cgi-bin/var.c [hunk 11]
```diff
@@ -234,6 +237,50 @@ cgiGetVariable(const char *name)	/* I - Name of variable */
...
+ *
+ * Returns 0 if the variable doesn't exist, 1 if it does
+ */
+
+int					/* O - 1 if variable exists */
+cgiArrayExists(const char *name,		/* I - Name of array variable */
+            int        element)		/* I - Element number (0 to N) */
+{
+  _cgi_var_t	*var;			/* Pointer to variable */
+
```
Comment (michaelrsweet) [inline cgi-bin/var.c:268]: two blank lines

File: cups/tls-darwin.c [hunk 12]
```diff
@@ -899,10 +899,8 @@ void
 _httpFreeCredentials(
     http_tls_credentials_t credentials)	/* I - Internal credentials */
 {
-  if (!credentials)
-    return;
-
-  CFRelease(credentials);
+  if (credentials)
```
Comment (michaelrsweet) [inline cups/tls-darwin.c:902]: separate unrelated changes

File: ppdc/ppdc-catalog.cxx [hunk 13]
```diff
@@ -107,11 +107,11 @@ ppdcCatalog::ppdcCatalog(const char *l,	// I - Locale
 
 	plist = CFPropertyListCreateWithStream(kCFAllocatorDefault, stream, 0, kCFPropertyListImmutable, NULL, NULL);
 
-	if (plist && CFGetTypeID(plist) == CFDictionaryGetTypeID())
-	  CFDictionaryApplyFunction((CFDictionaryRef)plist, (CFDictionaryApplierFunction)apple_add_message, this);
-
-	if (plist)
-	  CFRelease(plist);
+  if (plist) {
```
Comment (michaelrsweet) [inline ppdc/ppdc-catalog.cxx:110]: Separate unrelated changes.

Fix coding style/whitespace.

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
Please separate changes to backend/usb-darwin.c, cups/tls-darwin.c, and ppdc/ppdc-catalog.cxx as they are unrelated to the subject issue.

Also not sure why you are adding `void *` casts to the free calls - they aren't const pointers...

Some other issues (see comments) and you really don't need to explicitly free all of the cgiGetXxx values after use. Yes, it is "cleaner" but you free something that is allocated once and then you exit, which frees all memory used by the CGI anyways. Ideally the CGI code should use the libcups string pool instead of strdup and free, but given that this code will be going away in CUPS 3.x I don't see a need to go through a lot of churn.

**Comment by AtariDreams** [comment]
@michaelrsweet Done but now compilers are complaining about constness when doing free

**Comment by michaelrsweet** [comment]
> @michaelrsweet Done but now compilers are complaining about constness when doing free

That's the thing, the `(void *)` casts were added to non-const pointers.

**Pull request closed without merging**